### PR TITLE
[VL]Move getExtendedColumnarPostRules method into shim layer

### PR DESCRIPTION
**Pull request by JkSelf**
## What changes were proposed in this pull request?

Follow up https://github.com/oap-project/gluten/pull/3690 to resolve the [comment](https://github.com/oap-project/gluten/pull/3690#discussion_r1436860178).

## How was this patch tested?

Existing unit tests.

### CODE REVIEW

**Comment by JkSelf** [comment]
@zhztheplayer Can you help to review? Thanks.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/SparkPlanExecApiImpl.scala [hunk 1]
```diff
@@ -497,12 +496,16 @@ class SparkPlanExecApiImpl extends SparkPlanExecApi {
...
+
+    List(
+      spark =>
+        if (SparkShimLoader.getSparkShims.getExtendedColumnarPostRules(spark).nonEmpty) {
+          SparkShimLoader.getSparkShims.getExtendedColumnarPostRules(spark).get
+        } else {
+          new Rule[SparkPlan] {
+            override def apply(plan: SparkPlan): SparkPlan = plan
+          }
+        })
```
Comment (zhztheplayer) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/SparkPlanExecApiImpl.scala:508]: Could it be better with something like

```suggestion
SparkShimLoader.getSparkShims.getExtendedColumnarPostRules(spark) ::: List() 
```

Then in future we could add more rules that are common to all Spark versions into the rhs list.

**Review by zhztheplayer — COMMENTED** [review]
Thanks for improving this part of code!

File: gluten-core/src/main/scala/org/apache/spark/sql/execution/datasources/GlutenFormatWriterInjectsBase.scala [hunk 2]
```diff
@@ -66,4 +69,8 @@ trait GlutenFormatWriterInjectsBase extends GlutenFormatWriterInjects {
       transformStageCounter.incrementAndGet())
     FakeRowAdaptor(wst).execute()
   }
+
+  override def getExtendedColumnarPostRules(session: SparkSession): Rule[SparkPlan] = {
```
Comment (zhztheplayer) [inline gluten-core/src/main/scala/org/apache/spark/sql/execution/datasources/GlutenFormatWriterInjectsBase.scala:73]: nit: s/getExtendedColumnarPostRules/getExtendedColumnarPostRule

**Comment by JkSelf** [comment]
@zhztheplayer Resolved all your comments. Please help to review again. Thanks.

**Review by zhztheplayer — APPROVED** [review]
Thanks!

**Pull request merged**